run_examples.sh

```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
+
+build_dir=$1
+
+export TRACCC_TEST_DATA_DIR="$SCRIPT_DIR/data"
+
+
+set -e
+set -x
+
+$build_dir/bin/ccl_example $TRACCC_TEST_DATA_DIR/tml_pixels/event000000000-cells.csv
+$build_dir/bin/io_dec_par_example tml_detector/trackml-detector.csv tml_pixels 10
+$build_dir/bin/seq_example tml_detector/trackml-detector.csv tml_pixels 10
+$build_dir/bin/par_example tml_detector/trackml-detector.csv tml_pixels 10
+$build_dir/bin/seq_single_module
```

Comment:
:thinking: Instead of using a BASH-requiring script, I wonder if we shouldn't simply use [add_test(...)](https://cmake.org/cmake/help/latest/command/add_test.html) to declare all of these to [CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html). That would also have the upside of allowing the tests to run in parallel if desired. And would also avoid hardcoding the build directory layout (the `bin/` directory...) into the script.

Note that we could use a separate label for the "slow" tests/examples, if they turn out to run for much longer than what we'd like from unit tests. (So that we could explicitly filter them out/in depending on which tests we want to run.)

Reply:
Right, but we're running this only in the CI, it should run unmodified on macOS, and I'm not sure we should make decisions based on wanting to run on Windows, honestly.

I feel like adding them to the CTest test only makes sense if we are asserting the correctness of the examples, rather than just running them. Therefore I think this is the way to go until we potentially restructure what the example executables do / look like.

Reply:
I'm not sure I see the difference. :confused:

I actually really appreciate it when a project provides CTest recipes for how its executables could run. Yes, your script is a reasonable recipe as well, it's really just a matter of where we want to define that we want to run these as tests.

Since in the end you could also just fully forego this script, and just put these commands into the `builds.yml` file as well. It's purely a question of implementation how we want to do this exactly...

Reply:
@stephenswat can you break the tie? 😄

Reply:
I'm with Paul on this one, I don't personally think support for Windows should be a deciding factor in how we set this up, especially as the CI is virtually guaranteed to not be running on Windows.

Right now none of the examples are really in a spot where we can comprehensively test them for correctness, so that is something we can explore at a later point, I think. For now I don't see much value in anything other than simple running the examples and checking if they don't crash and burn.

Reply:
As another comment on Windows: I believe GitHub Action's Windows runners come with bash preinstalled.

Reply:
Fine. But then, do we need the script? Why don't we just extend `builds.yml`?

As for CTest: It literally just "runs something", which is supposed to return 0 on success, and something else on failure. Don't mix up CTest and GTest! So CTest is quite appropriate for testing whether something runs, or fails to run.

Reply:
I thought of one more argument. :stuck_out_tongue: If the "examples" are run by CTest, then we could easily see all of the failures at once.

Right now some of the tests are failing. So the examples are not even run. If everything would be run by CTest, we would see all the failures (possibly some of them from the examples), at once. :wink:

Reply:
Now that is a good point. :sweat_smile: